onnxruntime/core/common/threadpool.cc

```diff
@@ -21,12 +21,222 @@ limitations under the License.
 #include "core/common/eigen_common_wrapper.h"
 #include "core/platform/EigenNonBlockingThreadPool.h"
 #include "core/platform/ort_mutex.h"
+#ifdef _WIN32
+#include "processthreadsapi.h"
+#include <codecvt>
+#include <locale>
+#elif defined(__APPLE__)
+#include <cpuid.h>
+#else
+#include <sched.h>
+#endif
 
 namespace onnxruntime {
 
 namespace concurrency {
 
-// A sharded loop counter distributes loop iterations between a set of worker threads.  The iteration space of
+ThreadPoolProfiler::ThreadPoolProfiler(int num_threads, const CHAR_TYPE* threal_pool_name) : 
+    num_threads_(num_threads) {
+  child_thread_stats_.assign(num_threads, {});
+  if (threal_pool_name) {
+#ifdef _WIN32
+    using convert_type = std::codecvt_utf8<wchar_t>;
+    std::wstring_convert<convert_type, wchar_t> converter;
+    threal_pool_name_ = converter.to_bytes(threal_pool_name);
+#else
+    threal_pool_name_ = threal_pool_name;
+#endif
+  } else {
+    threal_pool_name_ = "unnamed_thread_pool";
+  }
+}
+
+ThreadPoolProfiler::~ThreadPoolProfiler() {
+  enabled_ = false;
+}
+
+void ThreadPoolProfiler::Start() {
+  enabled_ = true;
+}
+
+ThreadPoolProfiler::MainThreadStat& ThreadPoolProfiler::GetMainThreadStat() {
+  static thread_local std::unique_ptr<MainThreadStat> stat;
```

Comment:
so this should be fine if there is at most one ThreadPoolProfiler instance. if that's not the case, could do something like maintain a thread_local map from ThreadPoolProfiler* to MainThreadStat. the CUDA EP does something similar for per-thread state.
i think the "static" is not needed here - it would only affect the linkage but this is a function local variable. more details here: [https://en.cppreference.com/w/cpp/language/storage_duration](https://en.cppreference.com/w/cpp/language/storage_duration)